### ISSUE TITLE
Using the `viewport` stretch mode and resizing the window (or using "shrink" value other than 1) results in black screen if MSAA is forced on in NVIDIA or AMD settings

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot#18320]
**Godot version:** Git https://github.com/godotengine/godot/commit/7d6f210ccb5de9ef414f94ad42f9f3dea14c0493

**OS/device including version:** Fedora 27, NVIDIA 390.48

**Issue description:** When the window scaling mode is set to `2d` or `viewport`, setting the `display/window/stretch/shrink` project setting to a value other than 1 will cause the running project's rendering to break (it will appear to be frozen at the splash screen). If the window is resized by the user, it will turn into a black screen. No errors can be seen in the Debugger dock of the editor.

This also occurs when using the `viewport` stretch mode after resizing the window, even if `display/window/stretch/shrink` is set to 1.

**Steps to reproduce:** After making sure the window scaling mode is set to `2d` or `viewport`, set `display/window/stretch/shrink` to a value other than 1 in the Project Settings then run the project.

**Minimal reproduction project:**
*Resize the window **diagonally** after starting the project.*

- 4.x: [shrink_bug4.zip](https://github.com/user-attachments/files/20319948/shrink_bug4.zip)
- 3.x: [shrink_bug3.zip](https://github.com/user-attachments/files/20319949/shrink_bug3.zip)

___

**Update:** I can still reproduce this as of https://github.com/godotengine/godot/commit/6110bdee138febf3b04b47bc15b834bda7b99d52 (Fedora 30, NVIDIA 418.74). When the bug occurs, this appears in the console when Godot is started with `--verbose`:

```
ERROR: _gl_debug_print: GL ERROR: Source: OpenGL        Type: Error     ID: 1282        Severity: High  Message: GL_INVALID_OPERATION error generated. Source and destination dimensions must be identical with the current filtering modes.
   At: drivers/gles3/rasterizer_gles3.cpp:123.
```

**Comment by flashyincceo** [comment]
I am unable to replicate this bug on macOS 10.12.6 with the latest development version of Godot. 

I've tried setting the mode to 2d and viewport, and tried a couple of different `display/window/stretch/shrink` values, but the issues you are describing never popped up for me.

**Comment by akien-mga** [comment]
Works fine for me on 3.1.1-stable on Mageia 7 with both my graphic cards, KDE Plasma:
OpenGL ES 3.0 Renderer: Mesa DRI Intel(R) HD Graphics 630 (Kaby Lake GT2) 
OpenGL ES 3.0 Renderer: AMD VEGAM (DRM 3.30.0, 5.1.3-desktop-4.mga7, LLVM 8.0.0)

**Comment by daniel-abbott** [comment]
I've also been unable to replicate this behavior, using Windows 10 Pro, version 10.0.17134, neither in Godot 3.1.1-stable nor in 7d42ca8 and the GLES3 renderer:
OpenGL ES 3.0 Renderer: GeForce GTX 980M/PCIe/SSE2
OpenGL ES 3.0 Renderer: Intel(R) HD Graphics 530

**Comment by Calinou** [comment]
I just found out why this is occurring (and why nobody else has been able to reproduce it :stuck_out_tongue:).

This is caused by enabling a setting in the NVIDIA driver that tries to force MSAA in all applications:

![image](https://user-images.githubusercontent.com/180032/58114686-972e2380-7bf8-11e9-8f92-33da6b570b67.png)

(It generally doesn't work with newer applications, including Godot. However, it helps when running some old games or niche applications that don't feature MSAA settings.)

The Godot project setting that enables MSAA works fine with the `viewport` stretch mode otherwise:

![image](https://user-images.githubusercontent.com/180032/58114793-d0669380-7bf8-11e9-974c-c030dada92a4.png)

I'm not sure if we can fix this issue on our end. Someone with a NVIDIA graphics card on Windows should also try enable the equivalent setting and see if it causes the same bug.

**Comment by clayjohn** [comment]
Confirmed! :) Windows 10 NVidia GTX 1050 graphics card. When override MSAA is set it hangs on the splashscreen indefinitely and when resized turns black.

**Comment by RicardRC** [comment]
Windows 10 Nvidia GTX 1070 confirmed. Exact same sympthoms as clayjohnl.

**Comment by Calinou** [comment]
Are there any resources on fixing this (from the developer's side)? I'm not sure if reporting this issue to NVIDIA would be useful, as forcing MSAA is generally considered to be a legacy feature as it only impacts old applications.

**Comment by softgripper** [comment]
I just stumbled upon this bug too.

Disabling Antaliasing - Mode inside NVidia inspector "hid" the problem.

So the question is - is this a bug that won't get addressed (too hard to fix), or something that can't be fixed because of NVidia setting.

**Comment by clayjohn** [comment]
@softgripper It is an NVidia bug. :) If you want something done about it, you can report it to them. We can't fix NVidia software. 

Please do file a bug report with them though, the more people that report it, the more likely they are to fix the problem.

**Comment by softgripper** [comment]
I would, but I'm not clear on what the bug is.

Is it that MSAA is broken for all, or is it that Godot can't use NVidia's MSAA?

**Comment by clayjohn** [comment]
@softgripper NVidias MSAA override setting (in the driver menu) breaks window resizing in Windows 10.

**Comment by Calinou** [comment]
Interestingly, I ran into the same issue in [Quake3e](https://github.com/ec-/Quake3e) while testing a suggestion in [this issue](https://github.com/ioquake/ioq3/issues/422#issuecomment-612670825). So it seems Godot isn't the only software impacted by this bug.

**Comment by NoodleSushi** [comment]
I'm having the same issue for Godot v3.2.2.rc1.official GLES3. Whenever `shrink != 1`, Godot gets stuck on the splash screen. I changed my preferred graphics processor from "High-performance NVIDIA processor" to "Auto-select", and it doesn't get stuck anymore.

**Comment by amedley** [comment]
I know this is not ideal, but I just tried setting it to GLES2, and it works great! For my game, this is perfectly fine (we don't explicitly need GLES3), and I'm pretty sure for most devs using `shrink != 1` this is probably true. However in the case that someone really does need GLES3 and they are using `shrink != 1`, this is a huge issue, even just some way to detect on the client that this happens would help, because instructing the user that they need to modify their NVidia settings is virtually impossible (unless you bundle a windows specific executable that runs as a child process to check this ahead of time).

**Comment by Calinou** [comment]
clayjohn likely found the cause of this bug:

> When blitting a framebuffer the source and destination need to have the same MSAA setting. As far as I recall, the system_fbo always has an MSAA of 0. Therefore this will break whenever a user uses MSAA on their render target.
> —https://github.com/godotengine/godot/pull/35517#discussion_r370711192

**Comment by clayjohn** [comment]
@Calinou that comment only applies to the PR I commented on.

**Comment by Calinou** [comment]
> that comment only applies to the PR I commented on.

I know, but I have a feeling this omission is the cause of this bug. When forcing MSAA in the graphics driver, it will be forced on the "root" viewport/FBO/whatever instead of forcing it on individual viewports within the application. This is why it often doesn't work in modern engines.

**Comment by HeartoLazor** [comment]
I can't reproduce this issue with amd (7990 + 21.4.1 drivers).
Our internal QA has issues with a black screen + nvidia and I found this topic. There is a way to fix it from gdscript or editor settings?
I already had MSAA disabled in Godot editor.
We noticed with our team that the editor runs fine, but when you press play the game is not rendered. This doesn't make sense because the editor uses the same video driver as a godot game. 
Here is a image of the issue happening running a game from editor in a nvidia 1050, drivers 446.27, godot 3.3, GLES 3.0:
![image](https://user-images.githubusercontent.com/12376042/118904732-77fe2580-b8e8-11eb-8c2f-df10a1678c05.png)
Sound and input works, video not. Stay stuck at first frame (the splash screen). If the app is exported, same behavior but with black screen.
Notice the editor is fine at background, so why we can't apply the same editor behavior to games to fix this?.

**Comment by Listwon** [comment]
Official Godot 3.3 build, Windows 10 Nvidia GTX 1080, the same issue with forced MSAA and `Stretch mode: viewport`. One thing I noticed is that as long as viewport size and window size are the same it renders without issues. Resizing the window triggers black screen.

**Comment by Calinou** [comment]
> We noticed with our team that the editor runs fine, but when you press play the game is not rendered. This doesn't make sense because the editor uses the same video driver as a godot game.

The editor uses the `disabled` stretch mode whereas your project may be configured to use the `viewport` stretch mode instead.

**Comment by Miziziziz** [comment]
> There's no way to control driver-forced MSAA from the application. Detecting a MSAA override to warn the user about potential issues _might_ be possible, but I'm not sure.

Do you know if this is possible yet?

**Comment by Calinou** [comment]
> Do you know if this is possible yet?

Not that I know of. Ideally, we should fix whatever we're doing to be compliant with OpenGL specification, so that forced MSAA behaves correctly (or at the very least, does nothing).

**Comment by MagicHead333** [comment]
I was experiencing the same problem with the game Endoparacitic. I had the same exact symptoms as Calinou did and I seemed to have found a solution at least for my case. In the Nvidia Control Panel after turning off "Antialiasing - mode" that was previously set to "Override any application settings" It seemed to work for me. Windows 10, Nvidia RTX 3060. Sorry I'm new to this type of stuff

**Comment by Darenn** [comment]
Hey everyone!

I don't have the exact same issue but the solution given by @MagicHead333 solved mine.

2 out of the 5000 players who played my game reported this black screen issue (so it's rare). After some tests, turned out the culprit was the viewport stretch mode. Game ran ok with 2D mode or disabled. I wanted to post a proper issue but for some reason a minimal project with viewport mode worked great on these players computers.

I started looking what parameters in conjunction with the viewport mode could cause the issue. I did not find out, but using @MagicHead333 solution solved the problems for both players.

![image](https://github.com/godotengine/godot/assets/11272715/9834c20b-74d0-423d-bee7-62c10dcc6314)

It seems that this parameter is set right by default, hence why so few players had the problem.

Thank you @MagicHead333 !

**Comment by ashtonmeuser** [comment]
@Calinou Is any work being done to investigate this issue? It's nice having the features of GLES3 e.g. GPU particles, OpenGL `textureSize()` but this issue prevents a subset of users from even getting past the splash screen. It's not ideal to choose between supporting users and a modern dev experience.

**Comment by Calinou** [comment]
> Is any work being done to investigate this issue?

We don't know how to resolve this issue, as this is ultimately something triggered by the NVIDIA driver. NVIDIA is unlikely to fix the issue on their end, considering forcing MSAA is a legacy measure nowadays (the new NVIDIA app doesn't even expose this option anymore).

I suggest you keep a troubleshooting page somewhere on your store page so users are aware of this issue.

**Comment by ashtonmeuser** [comment]
@Calinou Appreciate the extra info. Sounds like it's not a high priority considering it only affects 3.x.

> When blitting a framebuffer the source and destination need to have the same MSAA setting. As far as I recall, the system_fbo always has an MSAA of 0. Therefore this will break whenever a user uses MSAA on their render target.

> When forcing MSAA in the graphics driver, it will be forced on the "root" viewport/FBO instead of forcing it on individual viewports within the application.

Perhaps I'm oversimplifying here, but do you think the following could be a viable stopgap solution?
1. Detect root viewport MSAA setting (possibly forced by Nvidia control center).
2. Apply the same setting to all child viewports.

This would ensure all viewports use the same MSAA setting thereby the discrepancy in source and destination mentioned by @clayjohn.

**Comment by ashtonmeuser** [comment]
I see. So the forced value may not be reflected by `get_tree().root.get_viewport().msaa`. Can we confirm this? I don't have the hardware to test otherwise I'd be more helpful.

**Comment by Calinou** [comment]
> So the forced value may not be reflected by get_tree().root.get_viewport().msaa. Can we confirm this? I don't have the hardware to test otherwise I'd be more helpful.

Indeed, the engine has no idea that the driver is overriding the root viewport's MSAA level (which in fact is not really the root viewport, as Godot always uses an intermediate framebuffer).

**Comment by KeyboardDanni** [comment]
Godot is already disabling nVidia's OpenGL "threaded optimization" through nvapi, essentially by reading/writing the application profile in the driver settings. Couldn't we do something similar with driver-level MSAA?

**Comment by Calinou** [comment]
> Godot is already disabling nVidia's OpenGL "threaded optimization" through nvapi, essentially by reading/writing the application profile in the driver settings. Couldn't we do something similar with driver-level MSAA?

We could do that as a workaround, but if possible, this should only apply when the viewport size doesn't match the window size.

**Comment by KeyboardDanni** [comment]
> We could do that as a workaround, but if possible, this should only apply when the viewport size doesn't match the window size (i.e. what happens with the `viewport` stretch mode or `shrink != 1.0`).

I don't think changing the application profile after the context is created would have any effect. Best bet is probably to have it as a project setting like with threaded optimization.

> 
> That said, for advanced users, this is nothing [NVIDIA Profile Inspector](https://github.com/Orbmu2k/nvidiaProfileInspector) can't override anyway 🙂

Correct me if I'm wrong, but isn't that tool just editing the same data that's already exposed by nvapi? And my understanding is that an application-specific profile will always override the global driver profile for any values changed from the defaults.

**Comment by graywes** [comment]
i updated my amd drivers recently and i opened godot for the first time and i was confused at why my project was just black when opened.



Attention, latest AMD drivers cause this bug too when anti aliasing is set to override application settings

**Comment by naturally-intelligent** [comment]
Having this problem with 4.4.1 export to Windows and Compatibility mode. Game runs, sound+music play, just black screen after initial splash screen.

Using "viewport" strech mode with 1.0 and fractional.

Giving the user option to set to "canvas_mode" as a command line argument does allow the game to be seen, but it is not the proper mode for the art to be displayed. Anyone else have a way to fix this?

UPDATE: Exporting from 4.5beta3 does resolve this for me!

**Comment by ashtonmeuser** [comment]
@Calinou @clayjohn Since the root cause of this issue seems to have been understood per #106267, can this fix be easily backported to 3.x? I am unfamiliar with the internals so if the fix is inapplicable, that's totally understandable. 

I know 3.x is a little long in the tooth, but this would be a great stability patch.

**Comment by clayjohn** [comment]
> [@Calinou](https://github.com/Calinou) [@clayjohn](https://github.com/clayjohn) Since the root cause of this issue seems to have been understood per [#106267](https://github.com/godotengine/godot/pull/106267), can this fix be easily backported to 3.x? I am unfamiliar with the internals so if the fix is inapplicable, that's totally understandable.
> 
> I know 3.x is a little long in the tooth, but this would be a great stability patch.

Yep, go ahead and prepare a backport if you want! Most of the code will be the same, but the 3.x version will be a little bit simpler.

**Comment by lawnjelly** [comment]
@ashtonmeuser let me know how you are getting on BTW, if you have stalled I can take a look at this soon-ish.

**Comment by ashtonmeuser** [comment]
> [@ashtonmeuser](https://github.com/ashtonmeuser) let me know how you are getting on BTW, if you have stalled I can take a look at this soon-ish.

Please feel free to take a look, @lawnjelly. After a brief effort, it seems this issue is a little too far in the Godot rendering internals for me. Further, without an NVIDIA card, I fear I wouldn't be able to properly test.
